### PR TITLE
Various improvements

### LINKED ISSUE #52: Can't load ROM files with special characters in filename.
If I try to load a ROM file with the word Pokémon in it, the core will fail to start up. If I rename the same file to say Pokemon then it will load normally. Standalone SameBoy can load the file even if it has a special character in its name.

### LINKED ISSUE #58: Better Auto System Model Support
Currently, the Auto System Model option only detects the difference between a Gameboy and Gameboy Color game, which it does by checking the game's file extension.
In the event it detects a Gameboy .gb game, it sets the model to be a monochrome DMG Gameboy.
This makes it so that you'd have to set the model manually in order for, for example, have SGB enhanced Gameboy games to display in color, which ultimately means a lot of overrides in the event of a large library.

A better would be to check the header of a Gameboy game and set the Auto System Model depending on whenever the SGB and/or CGB bits are set or not 
That way it would also work more like other Gameboy emulators in that regard.

### PR DESCRIPTION
**Pull request by jdgleaver**
This PR makes the following changes to the core:

- Direct file access has been replaced with VFS routines (libretro build only) (https://github.com/libretro/RetroArch/issues/12949)

- The core now sets `need_fullpath = false`, such that content is loaded from the frontend-supplied memory buffer. (https://github.com/libretro/RetroArch/issues/13005)

- Since the core no longer receives a content path, automatic ROM type detection cannot be based on file extension (`.gb`, `.gbc`). We now therefore inspect the ROM header, which means we can also distinguish SGB-enhanced content. As a result, the `System - Emulated Model` core option has been modified such that the `Auto` settings are:

    - `Auto Detect DMG/CGB` (default): Emulated hardware will be set to either Game Boy or Game Boy Color depending upon the ROM.
    - `Auto Detect DMG/SGB/CGB`: Emulated hardware will be set to either Game Boy, Super Game Boy or Game Boy Color depending upon the ROM.

A new `Auto Detected SGB Model` has also been added, to enable selection of which type of SGB hardware to emulate when content is detected as being SGB-enhanced.

- Since it was necessary to edit the core options anyway, these have been updated to v2 and option categories have been added.

Note that all changes here (apart from two `.gitignore` additions) are limited to the `libretro` directory - no 'emulator' code has been modified.

Closes  #58
Closes  #52 (since the core no longer loads ROMs directly)

### CODE REVIEW

**Comment by inactive123** [comment]
Hi @LIJI32, allowing you to review this as well before we hit the merge button on it. It's part of our Project IO effort as you can read in this blog post -

https://www.libretro.com/index.php/retroarch-1-9-11-released/

File: libretro/libretro.c [hunk 1]
```diff
@@ -1240,28 +1373,76 @@ void retro_run(void)
...
+
+    if (!data || (size < 0x146 + 1)) {
+        return ROM_TYPE_INVALID;
+    }
+
+    type     = ROM_TYPE_DMG;
+    cgb_flag = data[0x143];
+    sgb_flag = data[0x146];
+
+    if ((cgb_flag == 0x80) || (cgb_flag == 0xC0)) {
```
Comment (LIJI32) [inline libretro/libretro.c:1390]: `cgb_flag & 0x80` should be more technically correct

**Comment by LIJI32** [comment]
Other than my single comment on the CGB flag, the changes in SameBoy look fine to me. However I do find having the entire libretro VFS implementation committed into the repository a bit of a mess. Does the core even need the VFS and the compatibility layers if, after replacing SameBoy's `GB_load_rom` API with `GB_load_rom_from_buffer`, it not longer accesses the filesystem at all? Ideally, if still required, it should be dynamically linked, or be externally provided in a `.o` or `.a` form, so the only extra files needed are header files for exported APIs, or at the very least be a submodule/subrepo. The first solution makes it so it's no longer required to update and recompile the core whenever the VFS implementation changes, and the latter two at least remove the need to update the core's repository on such scenarios.

**Comment by LIJI32** [comment]
Additionally, maybe I'm missing something, but doesn't this change break compatibility with libretro frontends that are not aware of the new IO APIs?

**Comment by inactive123** [comment]
@LIJI32 No, we have designed it in a way so that it's perfectly backwards compatible.

**Comment by inactive123** [comment]
The libretro VFS files have to be committed into the repository. I disagree with this being a mess, and would argue that the opposite would have more disadvantages than advantages, and a submodule is not something I or jdgleaver would be in favor of here. We don't want to pull in the entire libretro-common for a small core like this, but only the basic required file. Besides, we have designed it in a way so that it's perfectly backwards compatible, so even if later VFS updates get pushed, the files committed as is will continue working.

**Comment by inactive123** [comment]
Given that there is no code disagreement then, I will go ahead and merge this.

**Pull request merged**